### ISSUE TITLE
🐛 Bug: Ersilia debug logs should limit printing model result

### ISSUE THREAD

**Issue by GemmaTuron** [ersilia-os/ersilia#1455]
### Describe the bug.

I am running Ersilia using the Python API, though I believe this is a problem on the CLI as well. When I run a model, it prints the outputs, but in the case of descriptors for example, which can have thousands of columns, this just clutters my terminal and makes it very difficult to follow. I am talking about this kind of printing. If the output is an array, only the first few values should be printed for clarity.

```
09:14:42 | DEBUG    | numeric_array
09:14:42 | DEBUG    | fingerprint
09:14:42 | DEBUG    | [-0.16217069327831268, -0.22873231768608093, 0.2539016902446747, 0.13426335155963898, 1.504109501838684, 0.4392009675502777, -0.023386826738715172, -0.0983983650803566, 0.11921375244855881, -0.07556186616420746, 1.3512521982192993, 0.122133769094944, 0.7075703144073486, 0.1358049362897873, 0.015197365544736385, 0.13897401094436646, -0.5125395655632019, 0.17096830904483795, -0.03985976800322533, -0.04970547556877136, 0.40511372685432434, -0.002354992786422372, -0.04410088434815407, -0.4040205776691437, -0.2317405790090561, 0.2253933995962143, -0.15340487658977509, -0.58453369140625, ...
```

### Operating environment

Ubuntu 24.04 LTS

**Comment by GemmaTuron** [comment]
Despite this the model seems to be working but I'd double check what is happening

**Comment by Abellegese** [comment]
Hey @GemmaTuron assign me on this task.

**Comment by Abellegese** [comment]
Hi @GemmaTuron how can i reproduce your bug?

**Comment by DhanshreeA** [comment]
Hey @Abellegese easiest way is to run a model like a descriptor such as eos7w6n, for example. The output from this model has >1000 columns which creates this issue.

**Comment by DhanshreeA** [comment]
@GemmaTuron @Abellegese I am removing Abel from this issue *for now*, because there are other tasks in the pipeline that Abel is taking up. Abel, let's come back to this after those.

**Comment by DhanshreeA** [comment]
@OlawumiSalaam I think this might be interesting for you to take up if you feel like it.

**Comment by Damisicode** [comment]
Can you help describe the process you followed @GemmaTuron maybe i can help

**Comment by DhanshreeA** [comment]
Hi @Damisicode thank you for your interest in contributing to Ersilia. Currently this issue is not a good starting point to contribute to this project. Please look out for issues labeled as good first issues. Thank you!

**Comment by Damisicode** [comment]
Alright, thanks.

please can you recommend base on the priority

**Comment by DhanshreeA** [comment]
@OlawumiSalaam is this something that interests you? Let me know!

**Comment by DhanshreeA** [comment]
Hey @OlawumiSalaam look at the following comment: https://github.com/ersilia-os/ersilia/issues/1455#issuecomment-2548408737

**Comment by OlawumiSalaam** [comment]
@DhanshreeA 
please can you help identify where in the Ersilia codebase these large outputs are being printed or logged (both CLI and Python API) so as to see the approach to fixing the issue.

**Comment by OlawumiSalaam** [comment]
@DhanshreeA 
I was able to reproduce the bug.
I think we can modify the output formatting

**Comment by DhanshreeA** [comment]
> [@DhanshreeA](https://github.com/DhanshreeA) please can you help identify where in the Ersilia codebase these large outputs are being printed or logged (both CLI and Python API) so as to see the approach to fixing the issue.

@OlawumiSalaam that's the task! ;)

**Comment by DhanshreeA** [comment]
> [@DhanshreeA](https://github.com/DhanshreeA) I was able to reproduce the bug. I think we can modify the output formatting

Perfect, please create a PR and test a couple of models: you can use eos4wt0, eos4u6p, eos7w6n, eos4avb for testing. Thanks!

**Comment by OlawumiSalaam** [comment]
> @DhanshreeA please can you help identify where in the Ersilia codebase these large outputs are being printed or logged (both CLI and Python API) so as to see the approach to fixing the issue.

**Comment by GemmaTuron** [comment]
> > [@DhanshreeA](https://github.com/DhanshreeA) I was able to reproduce the bug. I think we can modify the output formatting
> 
> Perfect, please create a PR and test a couple of models: you can use eos4wt0, eos4u6p, eos7w6n, eos4avb for testing. Thanks!

Hi also let's make sure the changes will not affect other models with short outputs, please test eos3b5e (1 column) eos69p9 (12 columns).

@OlawumiSalaam I believe what dhanshree is suggesting is that your work is to actually look into Ersilia's code and find the part that prints the outputs. It must be a logger somewhere. That is the difficult part of the task.

**Comment by OlawumiSalaam** [comment]
@GemmaTuron I am on it right away.

**Comment by DhanshreeA** [comment]
Yep, thanks Gemma! @OlawumiSalaam this task is a good entry level issue to get more familiar with the ersilia codebase and develop the intuition and debugging skills needed to pick up more difficult issues eventually. Please of course if you are stuck for more than a couple of hours, don't hesitate to ping me.

**Comment by OlawumiSalaam** [comment]
@DhanshreeA I was able to find the python file in the codebase but I am still working on the logic to fix the bug. here is the file I am working on
[https://github.com/OlawumiSalaam/ersilia/blob/master/ersilia/cli/commands/run.py](https://github.com/OlawumiSalaam/ersilia/blob/master/ersilia/cli/commands/run.py)

**Comment by DhanshreeA** [comment]
Hey @OlawumiSalaam I don't think this is the file, I neither see a log statement nor a print statement.

**Comment by OlawumiSalaam** [comment]
> Hey [@OlawumiSalaam](https://github.com/OlawumiSalaam) I don't think this is the file, I neither see a log statement nor a print statement.

[how about this part](https://github.com/OlawumiSalaam/ersilia/blob/5dc3e30f51e3dc93e5d88c0c0f7568f843633e46/ersilia/cli/commands/run.py#L72)
i thought it display results? I must have gotten it wrong?

**Comment by DhanshreeA** [comment]
@OlawumiSalaam my bad, that is indeed part of the code. 

```python
        if isinstance(result, types.GeneratorType):
            for result in mdl.run(input=input, output=output, batch_size=batch_size):
                if result is not None:
                    formatted = json.dumps(result, indent=4)
                    if as_table:
                        print_result_table(formatted)
                    else:
                        echo(formatted)
                else:
                    echo("Something went wrong", fg="red")
        else:
            if as_table:
                print_result_table(result)
            else:
                try:
                    echo(result)
                except:
                    print_result_table(result)
```

The `echo` command here is most definitely a culprit. It prints the entire result without regards for how big that result can be.

We should modify the input to this function and truncate it so we can print only a few dimensions. This will also need to be translated into the `print_result_table` when we get to it.

**Comment by GemmaTuron** [comment]
Hi @DhanshreeA and @OlawumiSalaam 

Sorry there seems to be a bit of a confusion here. We are talking about the Python API specifically, not the CLI commands.
If the user passes `ersilia run -i "COC1=CC23CCCN2CCC4=CC5=C(C=C4C3C1O)OCO5"` without specifing the output, the terminal SHOULD print all the output, because the user is asking for that.

The only thing I am referring to is when we run the python API and specify an output file (in my case, `.h5`) there is a DEBUG that still prints the output of the model, making the logs difficult to debug.

**Comment by GemmaTuron** [comment]
On a simple search for the `logger.debug` in Ersilia I found this lines in `sniff_bentoml.py`:

```
self.logger.debug(str(results)[:1000])
for r in results:
    if not r["output"]:
        raise EmptyOutputError(model_id=self.model_id, api_name=api_name)
```

The fix for this issue is simply to reduce the results to 10. I am sorry about this confusion but I do not agree that the output on the terminal should be modified, on the contrary. So I would backtrack on the open PR and just modify this.

**Comment by OlawumiSalaam** [comment]
@DhanshreeA 
Thank you for your feedback and guidance. I have implemented the changes you suggested regarding truncating the debug output, but unfortunately, the issue remains unresolved.

Here is what I did:

Created a branch for the fix:
```
git checkout master  
git checkout -b bugfix/truncate-debug-output 
``` 
Tested the current behavior:
I ran the following CLI commands to observe how the debug output behaves:
```
ersilia -v serve eos7w6n 
ersilia -v run -i "COC1=CC23CCCN2CCC4=CC5=C(C=C4C3C1O)OCO5"  
```

I observed that the debug logs were cluttered with long output, confirming the issue.
Modified the code:
In sniff_bentoml.py, I updated the logger.debug statement as follows:
`self.logger.debug(str(results[:10])) `
My intention was to limit the output to only the first 10 results.

Tested the new implementation:
After making the changes, I re-ran the CLI commands to check if the fix worked. Unfortunately, the debug output remains unchanged, and the cluttered logs are still being displayed.

This suggests that either the truncation logic isn’t being applied correctly or there might be another part of the code that’s generating the long outputs.

if you have any other suggestions or insights into what might be happening, I’d be happy to explore those further.

**Comment by DhanshreeA** [comment]
We've identified that this debug log specifically appears when using models not refactored with ersilia pack, because the flow of control in the code is different in both cases especially when trying serialize output formats that are not CSV. For this reason, we have decided to use eos7d58 to test this bugfix, because that is a model that's not been refactored with ersilia pack and it does have close to 100 columns. 

@OlawumiSalaam please use this and let us know what you find.

**Comment by GemmaTuron** [comment]
This line in `output.py` needs to be modified: `self.logger.info(f"Casted values: {vals}")` - I'd like to know though if this is a new addition and what do these casted values mean. If it is not a new addition, why wasn't it being printed before?

**Comment by GemmaTuron** [comment]
yes thanks @Abellegese 
I have realised I was in the bugfix_h5 branch still. This does not show up on the current master branch so I'll close this issue, sorry!

**Comment by Abellegese** [comment]
Yes go ahead.

**Comment by GemmaTuron** [comment]
Sorry @Abellegese 

Couple more questions on this. With the latest code in master, what prints is also very long, in two instances:
expanded output keys:

```
12:00:12 | DEBUG    | Data: outcome
12:00:12 | DEBUG    | Values: [-0.13663828, -0.26891047, -0.18683174, 0.10049557, 1.2305629, 0.31906578, 0.12811251, -0.6229073, 0
12:00:12 | DEBUG    | Getting pure dtype for outcome
12:00:12 | DEBUG    | This is the pure datatype: numeric_array
12:00:12 | DEBUG    | Datatype: numeric_array
12:00:12 | DEBUG    | Datatype has been matched: numeric_array over {'string_array', 'mixed_array', 'array', 'numeric_array'}
12:00:12 | DEBUG    | No merge key
12:00:12 | INFO     | Expanded output keys: ['dimension_0000', 'dimension_0001', 'dimension_0002', 'dimension_0003', 'dimension_0004', 'dimension_0005', 'dimension_0006', 'dimension_0007', 'dimension_0008', 'dimension_0009', 'dimension_0010', 'dimension_0011', 'dimension_0012', 'dimension_0013', 'dimension_0014', 'dimension_0015', 'dimension_0016', 'dimension_0017', 'dimension_0018', 'dimension_0019', 'dimension_0020', 'dimension_0021', 'dimension_0022', 'dimension_0023', 'dimension_0024', 'dimension_0025', 'dimension_0026', 'dimension_0027', 'dimension_0028', 'dimension_0029', 'dimension_0030', 'dimension_0031', 'dimension_0032', 'dimension_0033', 'dimension_0034', 'dimension_0035', 'dimension_0036', 'dimension_0037', 'dimension_0038', 'dimension_0039', 'dimension_0040', 'dimension_0041', 'dimension_0042', 'dimension_0043', 'dimension_0044', 'dimension_0045', 'dimension_0046', 'dimension_0047', 'dimension_0048', 'dimension_0049', 'dimension_0050', 'dimension_0051', 'dimension_0052', 'dimension_0053', 'dimension_0054', 'dimension_0055', 'dimension_0056', 'dimension_0057', 'dimension_0058', 'dimension_0059', 'dimension_0060', 'dimension_0061', 'dimension_0062', 'dimension_0063', 'dimension_0064', 'dimension_0065', 'dimension_0066', 'dimension_0067', 'dimension_0068', 'dimension_0069', 'dimension_0070', 'dimension_0071', 'dimension_0072', 'dimension_0073', 'dimension_0074'...]
``` 
and when the output keys are not expanded apparently? I am a bit lost as to what exactly this means
```
12:00:11 | DEBUG    | Schema available in /home/gturon/eos/dest/eos7w6n/api_schema.json
12:00:11 | DEBUG    | Time taken: 6.496260166168213
12:00:11 | DEBUG    | Done with unique posting
12:00:12 | WARNING  | Output key not expanded: val [[-0.13663828, -0.26891047, -0.18683174, 0.10049557, 1.2305629, 0.31906578, 0.12811251, -0.6229073, 0.12154026, 0.03959442, 0.7503495, 0.15416358, 0.67542857, 0.046326466, 0.3905704, 0.119286224, -1.0060033, 0.2002369, 0.107577436, 0.04336088, 0.071669415, 0.16342893, 0.0073369904, -0.34258473, -0.19075294, 0.2925139, -0.20740286, 0.32515246, -0.39340347, -0.12023362, 0.090462744, 0.5135876, -0.9351682, 0.5188422, 0.087276086, 0.09269708, -0.33887896, 0.42417476, 0.70925343, 0.42114967, 1.443135, -0.7115319, -0.2537332, -0.19113623, 0.019467216, -0.11281469, -0.38604265, -0.25951937, 0.75356555, -0.1158052, -0.119206496, 0.31423193, -0.31858176, 0.45985192, 0.0060250843, 0.020009896, 0.46494552, -0.8806706, 0.024950923, 0.4871387,...]
```

Do you know what I am referring to?

**Comment by Abellegese** [comment]
@GemmaTuron correct this is very annoying. I forgot to remove them before merge. Expanded keys are the column namea which are going to be in the output files header.


Just type "Expanded Key" in the vs code, you can find it this way. Or the file is in `ersilia/io/output.py` which the "annoying logs" are found.

**Comment by GemmaTuron** [comment]
Can we remove this prints or make them shorter? Maybe when you push a commit to Ersilia whenever it is that would be a good moment, if your remember to do it. No need to open a PR now until we fix the current issue with the workflows

**Comment by Abellegese** [comment]
Yes @GemmaTuron I don't actually need them anymore. I will remove them when this workflow things get fixed.

**Comment by GemmaTuron** [comment]
IS this issue good to be closed @Abellegese ?

**Comment by Abellegese** [comment]
Yes @GemmaTuron
